src/System.Text.Json/tests/Utf8JsonWriterTests.cs

```diff
@@ -575,6 +576,71 @@ public async Task UseAfterDisposeInvalidAsync(bool formatted, bool skipValidatio
             Assert.Throws<ObjectDisposedException>(() => jsonUtf8.Reset(output));
         }
 
+        [Fact]
+        public void FlushToStreamThrows_RemainsInConsistentState()
+        {
+            var stream = new ThrowingFromWriteMemoryStream();
+            var jsonUtf8 = new Utf8JsonWriter(stream);
+
+            // Write and flush some of an object.
+            jsonUtf8.WriteStartObject();
+            jsonUtf8.WriteString("someProp1", "someValue1");
+            jsonUtf8.Flush();
+
+            // Write some more, but fail while flushing to write to the underlying stream.
+            stream.ExceptionToThrow = new FormatException("uh oh");
+            jsonUtf8.WriteString("someProp2", "someValue2");
+            Assert.Same(stream.ExceptionToThrow, Assert.Throws<FormatException>(() => jsonUtf8.Flush()));
+
+            // Write some more.
+            stream.ExceptionToThrow = null;
+            jsonUtf8.WriteEndObject();
+
+            // Disposing should not fail.
+            jsonUtf8.Dispose();
+        }
+
+        [Fact]
+        public async Task FlushAsyncToStreamThrows_RemainsInConsistentState()
+        {
+            var stream = new ThrowingFromWriteMemoryStream();
+            var jsonUtf8 = new Utf8JsonWriter(stream);
+
+            // Write and flush some of an object.
+            jsonUtf8.WriteStartObject();
+            jsonUtf8.WriteString("someProp1", "someValue1");
+            await jsonUtf8.FlushAsync();
+
+            // Write some more, but fail while flushing to write to the underlying stream.
+            stream.ExceptionToThrow = new FormatException("uh oh");
+            jsonUtf8.WriteString("someProp2", "someValue2");
+            Assert.Same(stream.ExceptionToThrow, await Assert.ThrowsAsync<FormatException>(() => jsonUtf8.FlushAsync()));
+
+            // Write some more.
+            stream.ExceptionToThrow = null;
+            jsonUtf8.WriteEndObject();
+
+            // Disposing should not fail.
+            await jsonUtf8.DisposeAsync();
+        }
+
+        private sealed class ThrowingFromWriteMemoryStream : MemoryStream
+        {
+            public Exception ExceptionToThrow;
```

Comment:
nit: Why use a public field. Make it an auto-property instead?

Reply:
Because this is a helper class that exists purely to enable the test, and there's no need for the encapsulation that a property provides.  But I can change it.